neat_cache/lib/src/providers/resp.dart

```diff
@@ -0,0 +1,514 @@
+          'args',
+          'arguments for redis must be String, List<int>, int',
+        );
+      }
+
+      out.addByte(r'$'.codeUnitAt(0));
+      out.add(ascii.encode(bytes.length.toString()));
+      out.add(_newLine);
+      out.add(bytes);
+      out.add(_newLine);
+    }
+
+    final c = Completer<Object?>();
+    _pending.addLast(c);
+    try {
+      await _output.addStream(Stream.value(out.toBytes()));
+    } on Exception catch (e, st) {
+      await _abort(e, st);
+    }
+
+    return c.future;
+  }
+
+  /// Send `QUIT` command to redis and close the connection.
+  ///
+  /// If [force] is `true`, then the connection will be forcibly closed
+  /// immediately. Otherwise, connection will reject new commands, but wait for
+  /// existing commands to complete.
+  Future<void> close({bool force = false}) async {
+    _closing = true;
+
+    if (!_closing) {
+      // Always send QUIT message to be nice
+      try {
+        final quit = command(['QUIT']);
+        scheduleMicrotask(() async {
+          await quit.catchError((_) {/* ignore */});
+        });
+      } catch (_) {
+        // ignore
+      }
+    }
+
+    if (!force) {
+      scheduleMicrotask(() async {
+        await _output.close().catchError((_) {/* ignore */});
+      });
+      await _closed.future;
+    } else {
+      await _output.close().catchError((_) {/* ignore */});
+
+      // Resolve all outstanding requests
+      final pending = _pending.toList(growable: false);
+      _pending.clear();
+      final e = RedisConnectionException._('redis client forcibly closed');
+      final st = StackTrace.current;
+      pending.forEach((c) => c.completeError(e, st));
+    }
+    await _input.cancel();
+
+    assert(_pending.isEmpty, 'new pending requests added after close()');
+  }
+
+  /// Abort due to internal error
+  Future<void> _abort(Object e, StackTrace st) async {
+    if (!_closing) {
+      _log.warning('redis connection broken:', e, st);
+    }
+
+    _closing = true;
+
+    // Resolve all outstanding requests
+    final pending = _pending.toList(growable: false);
+    _pending.clear();
+    scheduleMicrotask(() {
+      pending.forEach((c) => c.completeError(e, st));
+    });
+
+    if (!_closed.isCompleted) {
+      _closed.complete();
+    }
+    await _input.cancel();
+
+    assert(_pending.isEmpty, 'new pending requests added after aborting');
+  }
+
+  void _readInput() async {
+    try {
+      while (true) {
+        Object? value;
+        try {
+          value = await _readValue();
+        } on RedisConnectionException catch (e, st) {
+          return await _abort(e, st);
+        }
+        if (_pending.isEmpty) {
+          return await _abort(
+            RedisConnectionException._('unexpected value from server'),
+            StackTrace.current,
+          );
+        }
+        final c = _pending.removeFirst();
+        if (value is RedisErrorException) {
+          // This is an error code returned by the server, it doesn't have a
```

Comment:
Wouldn't it be still useful if we'd keep the caller's stacktrace context?

Reply:
Maybe, but won't that get appended when you await the future?

Reply:
```dart
import 'dart:async';

Future<void> main() async {
  await _method();
}

Future<void> _method() async {
  final c = Completer();
  final f = c.future;
  c.completeError('', StackTrace.empty);
  await f;
}
```
It won't list neither the `main` nor the `_method` in the output. If you change to StackTrace.current, it will do both.